prov/lpp/src/lpp_atomic.c

```diff
@@ -0,0 +1,875 @@
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <stdlib.h>
+
+#include "lpp.h"
+
+struct lpp_atomic_spool_cbarg {
+	void *result_buf;
+};
+
+void lpp_atomic_handle_cmpl(Lpp_ep_t *lpp_epp, struct lpp_tx_entry *tx_entry,
+			    struct klpp_msg_hdr *hdr)
+{
+	struct lpp_atomic_hdr *atomic_hdr;
+	void *msg_data;
+
+	atomic_hdr = &tx_entry->atomic_hdr;
+
+	switch (atomic_hdr->mode) {
+	case LPP_ATOMIC:
+		lpp_completion_transmit(lpp_epp, lpp_epp->cntr_write,
+					tx_entry->flags,
+					tx_entry->msg_hdr.rcq_data,
+					tx_entry->context, hdr->status);
+		break;
+	case LPP_ATOMIC_FETCH:
+	case LPP_ATOMIC_CMP:
+		// "status" and "data_length" are a union. If data is present,
+		// we assume success and process the data. Otherwise, treat it
+		// as status.
+		if (!(hdr->flags & KLPP_MSG_DATA)) {
+			lpp_completion_transmit(lpp_epp, lpp_epp->cntr_write,
+						tx_entry->flags,
+						tx_entry->msg_hdr.rcq_data,
+						tx_entry->context, hdr->status);
+		} else {
+			msg_data = klpp_msg_data_offset(hdr, hdr->data_length);
+			memcpy(tx_entry->atomic_result.iov_base, msg_data,
+			       MIN(hdr->data_length,
+				   tx_entry->atomic_result.iov_len));
+			lpp_completion_transmit(lpp_epp, lpp_epp->cntr_write,
+						tx_entry->flags,
+						tx_entry->msg_hdr.rcq_data,
+						tx_entry->context, 0);
+		}
+		break;
+	default:
+		FI_WARN(&lpp_prov, FI_LOG_EP_DATA, "invalid atomic mode %u\n",
+			atomic_hdr->mode);
+	}
+}
+
+static void atomic_comp_spool_cb(Lpp_ep_t *lpp_epp, void *data, int status)
+{
+	struct lpp_atomic_spool_cbarg *cbarg = data;
+
+	free(cbarg->result_buf);
+}
+
+static void *atomic_res_buf_alloc(size_t size)
+{
+	void *buf;
+
+	// TODO: buffer pool?
+	if (posix_memalign(&buf, KLPP_MQ_CELL_SIZE, size) != 0) {
+		return NULL;
+	} else {
+		return buf;
+	}
```

Comment:
don't use `'{`  and `}` when the body has a single line. Same for other places.

Reply:
done